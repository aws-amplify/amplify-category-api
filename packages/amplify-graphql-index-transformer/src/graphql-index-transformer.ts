import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InvalidDirectiveError,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { IndexDirective } from '@aws-amplify/graphql-directives';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isListType, isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { generateKeyAndQueryNameForConfig, validateNotSelfReferencing } from './utils';

/**
 *
 */
export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];

  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', IndexDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as IndexDirectiveConfiguration,
      generateGetArgumentsInput(context.transformParameters),
    );

    /**
     * Impute Optional Fields
     * Start with sort key fields since both the name and queryField will work sort key fields into the generated name if necessary.
     */
    args.sortKeyFields = getOrGenerateDefaultSortKeyFields(args);
    args.name = getOrGenerateDefaultName(args);
    args.queryField = getOrGenerateDefaultQueryField(context, args);
    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

/**
 * Return the name if provided in our args, else
 * compute the name based on the field name, and sortKeyFields.
 */
const getOrGenerateDefaultName = (config: IndexDirectiveConfiguration): string => {
  const indexNameRegex = /^[A-Za-z0-9_\-\.]{3,255}$/;
  if (config.name) {
    if (!indexNameRegex.test(config.name)) {
      throw new Error(
        `The indexName is invalid. It should be between 3 and 255 characters. Only A–Z, a–z, 0–9, underscore characters, hyphens, and periods allowed.`,
      );
    }
    return config.name;
  }

  if (config.name === null) {
    throw new Error('Explicit null value not allowed for name field on @index');
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * Return the queryField if provided in our args, else
 * compute the queryField based on the field name, and sortKeyFields if the feature flag is enabled.
 */
const getOrGenerateDefaultQueryField = (
  context: TransformerSchemaVisitStepContextProvider,
  config: IndexDirectiveConfiguration,
): string | null => {
  const autoIndexQueryNamesIsEnabled = context.transformParameters.enableAutoIndexQueryNames;
  // Any explicit null will take effect, if enableAutoIndexQueryNames and no queryField is provide set to null for consistency
  if (config.queryField === null || (!autoIndexQueryNamesIsEnabled && !config.queryField)) {
    return null;
  }

  if (config.queryField) {
    return config.queryField;
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * sortKeyFields are optional so default to empty list,
 * if we get a raw object just wrap in an array,
 * else return the list which was provided correctly.
 */
const getOrGenerateDefaultSortKeyFields = (config: IndexDirectiveConfiguration): string[] => {
  if (!config.sortKeyFields) {
    return [];
  }
  if (!Array.isArray(config.sortKeyFields)) {
    return [config.sortKeyFields];
  }
  return config.sortKeyFields;
};

const validate = (config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void => {
  const { name, object, field, sortKeyFields } = config;
  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find((directive) => directive.name.value === 'model');

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${IndexDirective.name} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        const hasSortFields = peerDirective.arguments!.some(
          (arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0,
        );
        config.primaryKeyField = objectField;

        if (!hasSortFields && objectField.name.value === field.name.value) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === IndexDirective.name &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${IndexDirective.name} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }

    for (const peerDirective of objectField.directives!) {
      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0);

      if (
        !ctx.transformParameters.secondaryKeyAsGSI &&
        !hasSortFields &&
        objectField == config.primaryKeyField &&
        objectField.name.value === field.name.value
      ) {
        throw new InvalidDirectiveError(
          `Invalid @index '${name}'. You may not create an index where the partition key ` +
            'is the same as that of the primary key unless the index has a sort field. ' +
            'You cannot have a local secondary index without a sort key in the index.',
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums) || isListType(field.type)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums) || isListType(sortField.type)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
};
