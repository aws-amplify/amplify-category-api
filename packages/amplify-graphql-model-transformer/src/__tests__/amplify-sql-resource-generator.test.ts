import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import {
  MYSQL_DB_TYPE,
  constructDataSourceStrategies,
  getResourceNamesForStrategy,
  validateModelSchema,
} from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { SQLLambdaModelDataSourceStrategy } from '@aws-amplify/graphql-transformer-interfaces';
import { PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { ModelTransformer } from '../graphql-model-transformer';

describe('ModelTransformer with SQL data sources:', () => {
  const mysqlStrategy: SQLLambdaModelDataSourceStrategy = {
    name: 'mysqlStrategy',
    dbType: MYSQL_DB_TYPE,
    dbConnectionConfig: {
      usernameSsmPath: '/test/username',
      passwordSsmPath: '/test/password',
      hostnameSsmPath: '/test/hostname',
      databaseNameSsmPath: '/test/databaseName',
      portSsmPath: '/test/port',
    },
  };

  const validSchema = `
    type Post @model {
        id: ID! @primaryKey
        title: String!
    }
    type Comment @model {
      id: ID! @primaryKey
      content: String
    }
  `;

  it('should successfully transform simple valid schema', async () => {
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, mysqlStrategy),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(mysqlStrategy);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();
    expect(out.functions[`${resourceNames.sqlLambdaFunction}.zip`]).toBeDefined();
    expect(out.functions[`${resourceNames.sqlPatchingLambdaFunction}.zip`]).toBeDefined();
    validateModelSchema(parse(out.schema));
  });

  it('should assign SSM permissions', () => {
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, mysqlStrategy),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(mysqlStrategy);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();
    const [, policy] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaExecutionRolePolicy);
      }) || [];
    expect(policy).toBeDefined();
    const {
      Properties: { PolicyDocument },
    } = policy;
    expect(PolicyDocument.Statement.length).toEqual(2);

    const ssmStatements = PolicyDocument.Statement.filter(
      (statement: any) => JSON.stringify(statement.Action) === JSON.stringify(['ssm:GetParameter', 'ssm:GetParameters']),
    );
    expect(ssmStatements.length).toEqual(1);
    expect(ssmStatements[0].Resource).toEqual(
      Object.values(mysqlStrategy.dbConnectionConfig).map((ssmPath) => `arn:aws:ssm:*:*:parameter${ssmPath}`),
    );

    const secretsManagerStatements = PolicyDocument.Statement.filter(
      (statement: any) => statement.Action === 'secretsmanager:GetSecretValue',
    );
    expect(secretsManagerStatements.length).toEqual(0);

    const kmsStatements = PolicyDocument.Statement.filter((statement: any) => statement.Action === 'kms:Decrypt');
    expect(kmsStatements.length).toEqual(0);

    const cloudWatchStatements = PolicyDocument.Statement.filter(
      (statement: any) =>
        JSON.stringify(statement.Action) === JSON.stringify(['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']),
    );
    expect(cloudWatchStatements.length).toEqual(1);
    expect(cloudWatchStatements[0].Resource).toEqual('arn:aws:logs:*:*:*');

    expect(PolicyDocument).toMatchSnapshot();
  });

  it('should assign secrets manager permissions', () => {
    const secretArn = 'myfakesecretarn';
    const secretsManagerConfig = {
      ...mysqlStrategy,
      dbConnectionConfig: {
        databaseName: 'mydb',
        port: 3306,
        hostname: 'myfakehost',
        secretArn,
      },
    };
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, secretsManagerConfig),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(secretsManagerConfig);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();
    expect(sqlApiStack.Resources).toBeDefined();
    const [, policy] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaExecutionRolePolicy);
      }) || [];
    expect(policy).toBeDefined();

    const {
      Properties: { PolicyDocument },
    } = policy;
    expect(PolicyDocument.Statement.length).toEqual(2);

    const ssmStatements = PolicyDocument.Statement.filter(
      (statement: any) => JSON.stringify(statement.Action) === JSON.stringify(['ssm:GetParameter', 'ssm:GetParameters']),
    );
    expect(ssmStatements.length).toEqual(0);

    const secretsManagerStatements = PolicyDocument.Statement.filter(
      (statement: any) => statement.Action === 'secretsmanager:GetSecretValue',
    );
    expect(secretsManagerStatements.length).toEqual(1);
    expect(secretsManagerStatements[0].Resource).toEqual(secretArn);

    const kmsStatements = PolicyDocument.Statement.filter((statement: any) => statement.Action === 'kms:Decrypt');
    expect(kmsStatements.length).toEqual(0);

    const cloudWatchStatements = PolicyDocument.Statement.filter(
      (statement: any) =>
        JSON.stringify(statement.Action) === JSON.stringify(['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']),
    );
    expect(cloudWatchStatements.length).toEqual(1);
    expect(cloudWatchStatements[0].Resource).toEqual('arn:aws:logs:*:*:*');

    expect(PolicyDocument).toMatchSnapshot();
  });

  it('should assign secrets manager permissions with custom encryption key', () => {
    const secretArn = 'myfakesecretarn';
    const keyArn = 'myfakekeyarn';
    const secretsManagerConfig = {
      ...mysqlStrategy,
      dbConnectionConfig: {
        databaseName: 'mydb',
        port: 3306,
        hostname: 'myfakehost',
        secretArn,
        keyArn,
      },
    };
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, secretsManagerConfig),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(secretsManagerConfig);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();
    expect(sqlApiStack.Resources).toBeDefined();
    const [, policy] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaExecutionRolePolicy);
      }) || [];
    expect(policy).toBeDefined();
    const {
      Properties: { PolicyDocument },
    } = policy;
    expect(PolicyDocument.Statement.length).toEqual(3);

    const ssmStatements = PolicyDocument.Statement.filter(
      (statement: any) => JSON.stringify(statement.Action) === JSON.stringify(['ssm:GetParameter', 'ssm:GetParameters']),
    );
    expect(ssmStatements.length).toEqual(0);

    const secretsManagerStatements = PolicyDocument.Statement.filter(
      (statement: any) => statement.Action === 'secretsmanager:GetSecretValue',
    );
    expect(secretsManagerStatements.length).toEqual(1);
    expect(secretsManagerStatements[0].Resource).toEqual(secretArn);

    const kmsStatements = PolicyDocument.Statement.filter((statement: any) => statement.Action === 'kms:Decrypt');
    expect(kmsStatements.length).toEqual(1);
    expect(kmsStatements[0].Resource).toEqual(keyArn);

    const cloudWatchStatements = PolicyDocument.Statement.filter(
      (statement: any) =>
        JSON.stringify(statement.Action) === JSON.stringify(['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']),
    );
    expect(cloudWatchStatements.length).toEqual(1);
    expect(cloudWatchStatements[0].Resource).toEqual('arn:aws:logs:*:*:*');

    expect(PolicyDocument).toMatchSnapshot();
  });

  it('should process connection uri as input', () => {
    const connectionStringMySql = {
      ...mysqlStrategy,
      dbConnectionConfig: {
        connectionUriSsmPath: '/test/connectionUri',
      },
    };
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, connectionStringMySql),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(connectionStringMySql);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();

    // Check that SSM permissions are assigned
    const [, policy] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaExecutionRolePolicy);
      }) || [];

    expect(policy).toBeDefined();
    const {
      Properties: { PolicyDocument },
    } = policy;
    expect(PolicyDocument.Statement.length).toEqual(2);

    const ssmStatements = PolicyDocument.Statement.filter(
      (statement: any) => JSON.stringify(statement.Action) === JSON.stringify(['ssm:GetParameter', 'ssm:GetParameters']),
    );
    expect(ssmStatements.length).toEqual(1);
    expect(ssmStatements[0].Resource).toEqual(`arn:aws:ssm:*:*:parameter${connectionStringMySql.dbConnectionConfig.connectionUriSsmPath}`);

    // Check that secrets manager permissions are not assigned
    const secretsManagerStatements = PolicyDocument.Statement.filter(
      (statement: any) => statement.Action === 'secretsmanager:GetSecretValue',
    );
    expect(secretsManagerStatements.length).toEqual(0);

    const kmsStatements = PolicyDocument.Statement.filter((statement: any) => statement.Action === 'kms:Decrypt');
    expect(kmsStatements.length).toEqual(0);

    // Check that cloudwatch permissions are assigned
    const cloudWatchStatements = PolicyDocument.Statement.filter(
      (statement: any) =>
        JSON.stringify(statement.Action) === JSON.stringify(['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']),
    );
    expect(cloudWatchStatements.length).toEqual(1);
    expect(cloudWatchStatements[0].Resource).toEqual('arn:aws:logs:*:*:*');

    expect(PolicyDocument).toMatchSnapshot();

    // Check that the connection uri is passed to the lambda as an environment variable
    const [, sqlLambda] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaFunction);
      }) || [];
    expect(sqlLambda).toBeDefined();
    const envVars = sqlLambda.Properties.Environment.Variables;
    expect(envVars).toBeDefined();
    expect(envVars.connectionString).toBeDefined();
    expect(envVars.CREDENTIAL_STORAGE_METHOD).toEqual('SSM');
    expect(envVars.SSM_ENDPOINT).toBeDefined();
  });

  it('should accept multiple connection uris as input', () => {
    const connectionStringMySql = {
      ...mysqlStrategy,
      dbConnectionConfig: {
        connectionUriSsmPath: ['/test/connectionUri/1', '/test/connectionUri/2'],
      },
    };
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, connectionStringMySql),
    });
    expect(out).toBeDefined();
    const resourceNames = getResourceNamesForStrategy(connectionStringMySql);
    const sqlApiStack = out.stacks[resourceNames.sqlStack];
    expect(sqlApiStack).toBeDefined();

    // Check that SSM permissions are assigned
    const [, policy] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaExecutionRolePolicy);
      }) || [];

    expect(policy).toBeDefined();
    const {
      Properties: { PolicyDocument },
    } = policy;
    expect(PolicyDocument.Statement.length).toEqual(2);

    const ssmStatements = PolicyDocument.Statement.filter(
      (statement: any) => JSON.stringify(statement.Action) === JSON.stringify(['ssm:GetParameter', 'ssm:GetParameters']),
    );
    const ssmPaths = connectionStringMySql.dbConnectionConfig.connectionUriSsmPath;
    expect(ssmStatements.length).toEqual(1);
    expect(ssmStatements[0].Resource).toEqual(ssmPaths.map((ssmPath) => `arn:aws:ssm:*:*:parameter${ssmPath}`));

    // Check that secrets manager permissions are not assigned
    const secretsManagerStatements = PolicyDocument.Statement.filter(
      (statement: any) => statement.Action === 'secretsmanager:GetSecretValue',
    );
    expect(secretsManagerStatements.length).toEqual(0);

    const kmsStatements = PolicyDocument.Statement.filter((statement: any) => statement.Action === 'kms:Decrypt');
    expect(kmsStatements.length).toEqual(0);

    // Check that cloudwatch permissions are assigned
    const cloudWatchStatements = PolicyDocument.Statement.filter(
      (statement: any) =>
        JSON.stringify(statement.Action) === JSON.stringify(['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents']),
    );
    expect(cloudWatchStatements.length).toEqual(1);
    expect(cloudWatchStatements[0].Resource).toEqual('arn:aws:logs:*:*:*');

    expect(PolicyDocument).toMatchSnapshot();

    // Check that the connection uri is passed to the lambda as an environment variable
    const [, sqlLambda] =
      Object.entries(sqlApiStack.Resources!).find(([resourceName]) => {
        return resourceName.startsWith(resourceNames.sqlLambdaFunction);
      }) || [];
    expect(sqlLambda).toBeDefined();
    const envVars = sqlLambda.Properties.Environment.Variables;
    expect(envVars).toBeDefined();
    // The connection string information that is set in the lambda environment should be a valid JSON.
    expect(envVars.connectionString).toBeDefined();
    expect(JSON.parse(envVars.connectionString)).toEqual(ssmPaths);
    expect(envVars.CREDENTIAL_STORAGE_METHOD).toEqual('SSM');
    expect(envVars.SSM_ENDPOINT).toBeDefined();
  });
});
