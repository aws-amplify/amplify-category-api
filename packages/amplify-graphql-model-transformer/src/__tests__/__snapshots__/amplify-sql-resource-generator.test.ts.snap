// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelTransformer with SQL data sources: should assign SSM permissions 1`] = `
Object {
  "Statement": Array [
    Object {
      "Action": Array [
        "logs:CreateLogGroup",
        "logs:CreateLogStream",
        "logs:PutLogEvents",
      ],
      "Effect": "Allow",
      "Resource": "arn:aws:logs:*:*:*",
    },
    Object {
      "Action": Array [
        "ssm:GetParameter",
        "ssm:GetParameters",
      ],
      "Effect": "Allow",
      "Resource": Array [
        "arn:aws:ssm:*:*:parameter/test/username",
        "arn:aws:ssm:*:*:parameter/test/password",
        "arn:aws:ssm:*:*:parameter/test/hostname",
        "arn:aws:ssm:*:*:parameter/test/databaseName",
        "arn:aws:ssm:*:*:parameter/test/port",
      ],
    },
  ],
  "Version": "2012-10-17",
}
`;

exports[`ModelTransformer with SQL data sources: should assign secrets manager permissions 1`] = `
Object {
  "Statement": Array [
    Object {
      "Action": Array [
        "logs:CreateLogGroup",
        "logs:CreateLogStream",
        "logs:PutLogEvents",
      ],
      "Effect": "Allow",
      "Resource": "arn:aws:logs:*:*:*",
    },
    Object {
      "Action": "secretsmanager:GetSecretValue",
      "Effect": "Allow",
      "Resource": "myfakesecretarn",
    },
  ],
  "Version": "2012-10-17",
}
`;

exports[`ModelTransformer with SQL data sources: should assign secrets manager permissions with custom encryption key 1`] = `
Object {
  "Statement": Array [
    Object {
      "Action": Array [
        "logs:CreateLogGroup",
        "logs:CreateLogStream",
        "logs:PutLogEvents",
      ],
      "Effect": "Allow",
      "Resource": "arn:aws:logs:*:*:*",
    },
    Object {
      "Action": "secretsmanager:GetSecretValue",
      "Effect": "Allow",
      "Resource": "myfakesecretarn",
    },
    Object {
      "Action": "kms:Decrypt",
      "Effect": "Allow",
      "Resource": "myfakekeyarn",
    },
  ],
  "Version": "2012-10-17",
}
`;
